src/util/submit-addon.js

```diff
@@ -0,0 +1,344 @@
+
+import {createLogger} from './../util/logger.js';
+
+const log = createLogger(import.meta.url);
+
+export type SignResult = {|
+  id: string,
+  downloadedFiles: Array<string>,
+|};
+
+type ClientConstructorParams = {|
+  apiKey: string,
+  apiSecret: string,
+  apiHost: string,
+  apiJwtExpiresIn?: number,
+  validationCheckInterval?: number,
+  validationCheckTimeout?: number,
+  approvalCheckInterval?: number,
+  approvalCheckTimeout?: number,
+  downloadDir?: string,
+|};
+
+export default class Client {
+  apiKey: string;
+  apiSecret: string;
+  apiUrl: string;
+  apiJwtExpiresIn: number;
+  validationCheckInterval: number;
+  validationCheckTimeout: number;
+  approvalCheckInterval: number;
+  approvalCheckTimeout: number;
+  downloadDir: string;
+
+  constructor({
+    apiKey,
+    apiSecret,
+    apiHost,
+    apiJwtExpiresIn = 60 * 5, // 5 minutes
+    validationCheckInterval = 1000,
+    validationCheckTimeout = 300000, // 5 minutes.
+    approvalCheckInterval = 1000,
+    approvalCheckTimeout = 900000, // 15 minutes.
+    downloadDir = process.cwd(),
+  }: ClientConstructorParams) {
+    this.apiKey = apiKey;
+    this.apiSecret = apiSecret;
+    this.apiUrl = `${apiHost}/api/v5/addons/`;
+    this.apiJwtExpiresIn = apiJwtExpiresIn;
+    this.validationCheckInterval = validationCheckInterval;
+    this.validationCheckTimeout = validationCheckTimeout;
+    this.approvalCheckInterval = approvalCheckInterval;
+    this.approvalCheckTimeout = approvalCheckTimeout;
+    this.downloadDir = downloadDir;
+  }
+
+  fileFromSync(path: string): File {
+    return fileFromSync(path);
+  }
+
+  async doUploadSubmit(xpiPath: string, channel: string): Promise<string> {
+    const url = `${this.apiUrl}upload/`;
+    const formData = new FormData();
+    formData.set('channel', channel);
+    formData.set('upload', this.fileFromSync(xpiPath));
+    const { uuid } = await this.fetchJson(url, 'POST', formData);
+    return this.waitForValidation(uuid);
+  }
+
+  waitRetry(
+    successFunc: (detailResponseData: any) => string | null,
+    checkUrl: string,
+    checkInterval: number,
+    abortInterval: number,
+    context: string,
+  ): Promise<string> {
+    let checkTimeout;
+
+    return new Promise((resolve, reject) => {
+      const abortTimeout = setTimeout(() => {
+        clearTimeout(checkTimeout);
+        reject(new Error(`${context}: timeout.`));
+      }, abortInterval);
+
+      const pollStatus = async () => {
+        try {
+          const responseData = await this.fetchJson(
+            checkUrl, 'GET', undefined, 'Getting details failed.');
+
+          const success = successFunc(responseData);
+          if (success) {
+            resolve(success);
+          } else {
+            // Still in progress, so wait for a while and try again.
+            checkTimeout = setTimeout(pollStatus, checkInterval);
+          }
+        } catch (err) {
+          reject(err);
+        } finally {
+          clearTimeout(abortTimeout);
```

Comment:
While I was converting the tests from `nock` to `sinon.stub` the first validation timeout test that I converted was entering in an infinite look and never completing, I'm not sure why the same test wasn't failing in a similar way in its `nock` form (I haven't looked too deeply into that) but after looking to this method I wondered if `abortTimeout` was not supposed to be cleared when we are entering in the `else` branch at line 102 to keep polling and instead to be only cleared if we have done with it or we are giving up (and so only if we are resolving or rejecting the wrapping promise).

Given that the default timeout seems to be 5min I'd guess it is not meant to be a timeout for a single request, and instead that is was meant to be a timeout for the entire time we are willing to keep polling for the validation result.

If that is the case, then I think that `clearTimeout` should be called right before the resolve and reject calls and the `finally` block removed (as I've tweaked it in the commit I linked before, https://github.com/rpl/web-ext/commit/c192cef1454c11f49a297758e3bad377152119a5 to make it pass all validation timeout tests rewritten from `nock` to `sinon.stub`).

@eviljeff wdyt? does what I describe above match the intention of this method? (and the proposed tweak a reasonable fix for the infinite look that the test case rewritten with sinon.stub seems to be able to trigger)
or should the timeout we cleared here as in this version of the method but we should tweak it to stop retrying after a certain number of retries?

Reply:
This was also a requested change from a few weeks ago but you've identified the flaw - before the `clearTimeout` was moved to a `finally`, it wasn't called when the wait was still in progress.  It's supposed to be an overall timeout on the process so it's not supposed to be clear on each check iteration.

Maybe it would be more sensible to have a checkInterval time setting and a numberOfChecks setting rather than two times.  That would be incompatible with the settings used with signAddon though.